src/main/scala/com/fulcrumgenomics/umi/VanillaUmiConsensusCaller.scala

```diff
@@ -0,0 +1,367 @@
+    val secondOfPairConsensus = consensusFromSamRecords(records=secondOfPair)
+    (firstOfPairConsensus, secondOfPairConsensus) match {
+      case (None, None)                                       => // reject
+        rejectRecords(records=firstOfPair ++ secondOfPair)
+      case (Some(_), None) | (None, Some(_)) if needBothPairs => // reject
+        rejectRecords(records=firstOfPair ++ secondOfPair)
+      case (firstRead, secondRead)                            => // output
+        this.createAndEnqueueSamRecordPair(firstRecords=firstOfPair, firstRead=firstRead, secondRecords=secondOfPair, secondRead=secondRead)
+        success = true
+    }
+
+    if (success) true // consensus created
+    else if (this.iter.isEmpty) false // no more records, don't try again
+    else this.advance() // no consensus, but more records, try again
+  }
+
+  private def rejectRecords(records: Seq[SAMRecord]): Unit = this.rejects.foreach(rej => records.foreach(rej.addAlignment))
+
+  /** Adds a SAM record from the underlying iterator to the buffer if either the buffer is empty or the SAM tag is
+    * the same for the records in the buffer as the next record in the input iterator.  Returns true if a record was
+    * added, false otherwise.
+    */
+  private def nextGroupOfRecords(): List[SAMRecord] = {
+    if (this.iter.isEmpty) Nil
+    else {
+      val tagToMatch = this.iter.head.getStringAttribute(options.tag)
+      val buffer = ListBuffer[SAMRecord]()
+      while (this.iter.hasNext && this.iter.head.getStringAttribute(options.tag) == tagToMatch) {
+        val rec = this.iter.next()
+        buffer += rec
+      }
+      progress.map(_.record(buffer:_*))
+      buffer.toList
+    }
+  }
+
+  /** Split records into those that should make a single-end consensus read, first of pair consensus read,
+    * and second of pair consensus read, respectively.  The default method is to use the SAM flag to find
+    * unpaired reads, first of pair reads, and second of pair reads.
+    */
+  protected def subGroupRecords(records: Seq[SAMRecord]): (Seq[SAMRecord], Seq[SAMRecord],Seq[SAMRecord]) = {
+    val fragments    = records.filter { rec => !rec.getReadPairedFlag }
+    val firstOfPair  = records.filter { rec => rec.getReadPairedFlag && rec.getFirstOfPairFlag }
+    val secondOfPair = records.filter { rec => rec.getReadPairedFlag && rec.getSecondOfPairFlag }
+    (fragments, firstOfPair, secondOfPair)
+  }
+
+  /** Returns the next read name with format "<prefix>:<idx>", where "<prefix>" is either the supplied prefix or the
+    * longest common prefix of all read names, and "<idx>" is the 1-based consensus read index.  If no prefix was found,
+    * "CONSENSUS" is used.  If no records are given, the empty string is returned.
+    */
+  private def nextReadName(records: Seq[SAMRecord]): String = {
+    if (records.isEmpty) ""
+    else this.options.tag + ":" + records.head.getStringAttribute(this.options.tag)
```

Comment:
I am not sure the read name is going to be unique!

Also, the doc for this method needs updating.